File: scripts/program_transactions/mint_nft.ts
// Mint NFT
import * as anchor from '@coral-xyz/anchor';
import {
  Keypair,
  PublicKey,
  Connection,
  clusterApiUrl,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
  ComputeBudgetProgram,
} from '@solana/web3.js';
import {
  TOKEN_PROGRAM_ID,
  getAssociatedTokenAddress,
  ASSOCIATED_TOKEN_PROGRAM_ID,
} from '@solana/spl-token';
import { getPayerKeypair } from '../../utils/utils';
import idl from '../../target/idl/collectible_vault.json'; // Import the IDL JSON
import { CollectibleVault } from '../../target/types/collectible_vault'; // Import TypeScript types
import { COLLECTION_COUNTER, COLLECTION_MINT, METADATA_PROGRAM_ID } from '../constants';

const payerKeypair = getPayerKeypair();

// Solana connection
const connection = new Connection(clusterApiUrl('devnet'), 'confirmed');
const provider = new anchor.AnchorProvider(connection, new anchor.Wallet(payerKeypair), {
  preflightCommitment: 'confirmed',
});
anchor.setProvider(provider);

const program = new anchor.Program<CollectibleVault>(idl as CollectibleVault, provider);

// Creates Metaplex NFT Collection from anchor instruction.
async function createNft() {
  await checkCollection();
  console.log('Creating individual NFT...');

  // Add compute budget instruction
  const modifyComputeUnits = ComputeBudgetProgram.setComputeUnitLimit({
    units: 300000
  });

  // Generate a new mint keypair for the collection
  const mint = Keypair.generate();

  // Derive PDA for metadata account
  const [metadataPDA] = PublicKey.findProgramAddressSync(
    [
      Buffer.from('metadata'),
      new PublicKey(
        idl.instructions
          .find((i) => i.name === 'mint_nft')
          ?.accounts.find((a) => a.name === 'token_metadata_program')?.address!
      ).toBuffer(),
      mint.publicKey.toBuffer(),
    ],
    new PublicKey(
      idl.instructions
        .find((i) => i.name === 'mint_nft')
        ?.accounts.find((a) => a.name === 'token_metadata_program')?.address!
    )
  );

  // Derive PDA for master edition account
  const [masterEditionPDA] = PublicKey.findProgramAddressSync(
    [
      Buffer.from('metadata'),
      METADATA_PROGRAM_ID.toBuffer(),
      mint.publicKey.toBuffer(),
      Buffer.from('edition'),
    ],
    METADATA_PROGRAM_ID
  );

  // Derive collection metadata PDA
  const [collectionMetadataPDA] = PublicKey.findProgramAddressSync(
    [Buffer.from('metadata'), METADATA_PROGRAM_ID.toBuffer(), COLLECTION_MINT.toBuffer()],
    METADATA_PROGRAM_ID
  );

  // Derive collection master edition PDA
  const [collectionMasterEditionPDA] = PublicKey.findProgramAddressSync(
    [
      Buffer.from('metadata'),
      METADATA_PROGRAM_ID.toBuffer(),
      COLLECTION_MINT.toBuffer(),
      Buffer.from('edition'),
    ],
    METADATA_PROGRAM_ID
  );

  // Derive or use a dummy for collection authority record
  const [collectionAuthorityRecord] = PublicKey.findProgramAddressSync(
    [
      Buffer.from('metadata'),
      METADATA_PROGRAM_ID.toBuffer(),
      COLLECTION_MINT.toBuffer(),
      Buffer.from('collection_authority'),
      payerKeypair.publicKey.toBuffer(),
    ],
    METADATA_PROGRAM_ID
  );

  // Get the associated token account for the payer
  const tokenAccount = await getAssociatedTokenAddress(mint.publicKey, payerKeypair.publicKey);

  console.log('Minting with accounts:');
  console.log(`- Mint: ${mint.publicKey.toString()}`);
  console.log(`- Metadata: ${metadataPDA.toString()}`);
  console.log(`- Master Edition: ${masterEditionPDA.toString()}`);
  console.log(`- Collection Mint: ${COLLECTION_MINT.toString()}`);
  console.log(`- Collection Metadata: ${collectionMetadataPDA.toString()}`);
  console.log(`- Collection Counter: ${COLLECTION_COUNTER.toString()}`);
  const accounts = {
    mint: mint.publicKey,
    metadata: metadataPDA,
    masterEdition: masterEditionPDA,
    tokenAccount: tokenAccount,
    payer: payerKeypair.publicKey,
    systemProgram: SystemProgram.programId,
    tokenProgram: TOKEN_PROGRAM_ID,
    associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
    rent: SYSVAR_RENT_PUBKEY,
    tokenMetadataProgram: METADATA_PROGRAM_ID,
    collection: collectionMetadataPDA, // This is the same as collectionMetadata if you refactored
    collectionMint: COLLECTION_MINT,
    collectionMetadata: collectionMetadataPDA,
    collectionMasterEdition: collectionMasterEditionPDA,
    collectionCounter: COLLECTION_COUNTER,
    // collectionAuthorityRecord: collectionAuthorityRecord,
  };

  const tx = await program.methods
    .mintNft('https://metadata.y00ts.com/y/6869.json')
    .accounts(accounts)
    .preInstructions([modifyComputeUnits])
    .signers([payerKeypair, mint])
    .rpc();

  console.log(`✅ Individual NFT Created! TX: ${tx}`);
  console.log(`NFT Mint Address: ${mint.publicKey.toBase58()}`);
}

async function checkCollection() {
  try {
    const accountInfo = await connection.getAccountInfo(COLLECTION_MINT);
    console.log('Collection exists:', accountInfo !== null);

    if (accountInfo) {
      // Instead, derive the metadata PDA and check that
      const [collectionMetadataPDA] = PublicKey.findProgramAddressSync(
        [Buffer.from('metadata'), METADATA_PROGRAM_ID.toBuffer(), COLLECTION_MINT.toBuffer()],
        METADATA_PROGRAM_ID
      );

      const metadataAccountInfo = await connection.getAccountInfo(collectionMetadataPDA);
      console.log('Collection metadata exists:', metadataAccountInfo !== null);

      // Also check the master edition
      const [collectionMasterEditionPDA] = PublicKey.findProgramAddressSync(
        [
          Buffer.from('metadata'),
          METADATA_PROGRAM_ID.toBuffer(),
          COLLECTION_MINT.toBuffer(),
          Buffer.from('edition'),
        ],
        METADATA_PROGRAM_ID
      );

      const masterEditionAccountInfo = await connection.getAccountInfo(collectionMasterEditionPDA);
      console.log('Collection master edition exists:', masterEditionAccountInfo !== null);
    }
  } catch (error) {
    console.error('Error checking collection:', error);
  }
}

createNft().catch(console.error);

File: programs/collectible_vault/src/instructions/mint_nft.rs
use anchor_lang::prelude::*;
use anchor_spl::{
    token::{Mint, Token, TokenAccount},
    associated_token::AssociatedToken,
};
use crate::{
    state::CollectionCounter,
    errors,
    constants::get_admin_account_pubkey,
};

#[derive(Accounts)]
pub struct MintNFT<'info> {
    #[account(
        init,
        payer = payer,
        mint::decimals = 0,
        mint::authority = payer.key(),
        mint::freeze_authority = payer.key(),
    )]
    pub mint: Account<'info, Mint>,

    // Metadata account
    /// CHECK: Metaplex metadata validation handles this.
    #[account(mut)]
    pub metadata: UncheckedAccount<'info>,

    /// CHECK: Metaplex master edition validation handles this.
    #[account(mut)]
    pub master_edition: UncheckedAccount<'info>,

    // User token account
    #[account(
        init_if_needed,
        payer = payer,
        associated_token::mint = mint,
        associated_token::authority = payer,
    )]
    pub token_account: Account<'info, TokenAccount>,

    #[account(mut, constraint = payer.key() == get_admin_account_pubkey() @ errors::ErrorCode::UnauthorizedTransactionSigner)]
    // Payer account
    pub payer: Signer<'info>,
    // System programs
    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub rent: Sysvar<'info, Rent>,

    //Metaplex program
    /// CHECK: Metaplex program ID check
    #[account(address = mpl_token_metadata::ID)]
    pub token_metadata_program: UncheckedAccount<'info>,

    // Metaplex collection that already exists
    #[account(mut)]
    /// CHECK: Metaplex will validate the collection account
    pub collection: UncheckedAccount<'info>,

    // Collection accounts
    /// CHECK: Collection mint
    pub collection_mint: Account<'info, Mint>,

    /// CHECK: Collection metadata account
    #[account(mut)]
    pub collection_metadata: UncheckedAccount<'info>,

    /// CHECK: Collection master edition
    pub collection_master_edition: UncheckedAccount<'info>,

    #[account(
        mut,
        seeds = [b"collection_counter", collection_mint.key().as_ref()],
        bump,
        constraint = collection_counter.collection_mint == collection_mint.key() @ errors::ErrorCode::CollectionMintDoesNotMatch
    )]
    pub collection_counter: Account<'info, CollectionCounter>,
}